File: LeetCode/sword/25.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* mergeTwoLists(ListNode *l1, ListNode *l2) {
		ListNode *head1 = l1;
		ListNode *head2 = l2;
		if(l1 == nullptr) {
			return l2;
		}else if(l2 == nullptr) {
			return l1;
		}
		ListNode *head = nullptr, *tail = nullptr;
		while(head1 != nullptr && head2 != nullptr) {
			if(head1->val < head2->val) {
				if(tail == nullptr) {
					head = tail = head1;
				}else {
					tail->next = head1;
					tail = head1;
				}
				head1 = head1->next;
				tail->next = nullptr;
			}else {
				if(tail == nullptr) {
					head = tail = head2;
				}else {
					tail->next  =  head2;
					tail = head2;
				}
				head2 = head2->next;
				tail->next = nullptr;
			}
		}
		if(head1 != nullptr) {
			tail->next = head1;
		}else if(head2 != nullptr) {
			tail->next = head2;
		}
		return head;
	}
};

File: LeetCode/25.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        res = root = ListNode(-1)
        root.next = head
        tmp = head
        cnt = 0
        while(tmp != None):
            tmp = tmp.next
            cnt += 1
        while(cnt >= k):
            for i in range(k-1):
                node = root.next
                root.next = head.next
                head.next = root.next.next
                root.next.next = node
            root = head
            head = head.next
            cnt -= k
        return res.next